api/query_test.go

```diff
@@ -36,6 +38,15 @@ type fenceTest struct {
 	expectedError  string
 }
 
+type metadataTest struct {
+	name           string
+	method         string
+	metadata       testMetadataQuery
+	metadataQuery  string
```

Comment:
I really dislike that you have this "either/or" design. It introduce too many questions and ambiguity. What's the rules for when to use witch? What purpose do they serve (one vs the other)? What happens if I specify both?

Reply:
Purpose answered in another comment.
idea was to use struct where possible for nicer readability and jsonQuery where desired json can't be represented as a struct.

Idea is that Query takes precedence over struct as it is more "final" form.

But let me know, if you want to combine everything to json, without structs at all, it is doable. Code would probably be less readable, but it might bring more additional benefits as getting rid of all the testQuery structs you despise so!

Reply:
> idea was to use struct where possible for nicer readability and jsonQuery where desired json can't be represented as a struct.

I would solve that to have different tests for "invalid / incomplete json" cases and other tests.

> But let me know, if you want to combine everything to json, without structs at all, it is doable. Code would probably be less readable

No thank you 😄 
> but it might bring more additional benefits as getting rid of all the testQuery structs you despise so!
See my other comment, you don't _need_ them in the first place